integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -407,49 +405,25 @@ private class OuterFuture<T>(jobToCancel: Job): ListenableFuture<T> {
         return innerFuture.isDone
     }
 
-    override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
-        return innerFuture.cancel(mayInterruptIfRunning)
-    }
-}
-
-/**
- * Serves as a state machine for [Future] cancellation.
- *
- * If this future gets [cancelled][Future.cancel], [jobToCancel] will be [cancelled][Job.cancel] as well.
- *
- * [AbstractFuture] has a highly-correct atomic implementation of `Future`'s completion and
- * cancellation semantics. By using that type, the [OuterFuture] can delegate its semantics to
- * _this_ `Future` `get()` the result in such a way that the `Deferred` is always complete when
- * returned.
- */
-private class InnerFuture<T>(
-    private val jobToCancel: Job
-) : AbstractFuture<Result<T>>() {
-
-    /**
-     * When the attached coroutine [isCompleted][Job.isCompleted], its [Result] should be passed to this method.
-     *
-     * This should succeed barring a race with external cancellation.
-     */
-    fun complete(result: Result<T>) = set(result)
-
     /**
      * Tries to cancel [jobToCancel] if `this` future was cancelled. This is fundamentally racy.
      *
-     * The call to `cancel()` will try to cancel this Future: if and only if cancellation of this
+     * The call to `cancel()` will try to cancel [innerFuture]: if and only if cancellation of [innerFuture]
      * succeeds, [jobToCancel] will have its [Job.cancel] called.
      *
      * This arrangement means that [jobToCancel] _might not successfully cancel_, if the race resolves
      * in a particular way. [jobToCancel] may also be in its "cancelling" state while this
      * ListenableFuture is complete and cancelled.
-     *
-     * [OuterFuture] collaborates with this class to present a more cohesive picture and ensure
-     * that certain combinations of cancelled/cancelling states can't be observed.
      */
-    @Suppress("UnstableApiUsage")
-    override fun afterDone() {
-        if (isCancelled) {
+    override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
+        // TODO: call jobToCancel.cancel() _before_ running the listeners.
+        //  Future.cancel may execute all the listeners which can delay cancellation of jobToCancel.
```

Comment:
"innerFuture.cancel() will execute innerFuture's listeners. This delays cancellation of jobToCancel until after innerFuture's listeners have already run." 

Something like that?

The distinction between "innerFuture" and "this Future" is a bit squishy. Maybe naming innerFuture "delegate" or something would be clearer. Not sure.

Reply:
Fixed the comment.

It's not exactly a "delegate", right? Renamed `innerFuture` to `auxFuture` and `OuterFuture` to `JobListenableFuture`. Let me know if you can think of a better naming.

Reply:
No, I don't have a better name to hand. "innerFuture", "actualFuture" or "internalFuture", anything like that...

The inner Future is the concrete implementation, and the outer Future is a proxy type that both owns the reference and invokes methods on it. I haven't read the Gang of Four in long enough to describe this design pattern better.